Tell the window manager when a click should activate the window: on press, or on release for draggable items, so dragging from a background window doesn't raise it.

diff --git a/gdk/x11/gdkwindow-x11.c b/gdk/x11/gdkwindow-x11.c
index c60cecd..3d7595a 100644
--- a/gdk/x11/gdkwindow-x11.c
+++ b/gdk/x11/gdkwindow-x11.c
@@ -795,12 +795,13 @@ static void
 set_wm_protocols (GdkWindow *window)
 {
   GdkDisplay *display = gdk_window_get_display (window);
-  Atom protocols[4];
+  Atom protocols[5];
   int n = 0;
   
   protocols[n++] = gdk_x11_get_xatom_by_name_for_display (display, "WM_DELETE_WINDOW");
   protocols[n++] = gdk_x11_get_xatom_by_name_for_display (display, "WM_TAKE_FOCUS");
   protocols[n++] = gdk_x11_get_xatom_by_name_for_display (display, "_NET_WM_PING");
+  protocols[n++] = gdk_x11_get_xatom_by_name_for_display (display, "_GTK_ACTIVATE_ON_CLICK");
 
 #ifdef HAVE_XSYNC
   if (GDK_X11_DISPLAY (display)->use_sync)
diff --git a/gtk/gtkdnd.c b/gtk/gtkdnd.c
index d2bc74f..177b436 100644
--- a/gtk/gtkdnd.c
+++ b/gtk/gtkdnd.c
@@ -59,6 +59,7 @@
 #include "gtkintl.h"
 #include "gtkmain.h"
 #include "gtkplug.h"
+#include "gtkprivate.h"
 #include "gtktooltipprivate.h"
 #include "gtkwindow.h"
 #include "gtkrender.h"
@@ -1728,6 +1729,8 @@ gtk_drag_begin_internal (GtkWidget           *widget,
   GdkAtom selection;
   gboolean managed;
 
+  _gtk_main_cancel_activate_on_release ();
+
   managed = gtk_drag_is_managed (widget);
 
   pointer = keyboard = NULL;
diff --git a/gtk/gtkdndprivate.h b/gtk/gtkdndprivate.h
index 89f1b0f..7dc420f 100644
--- a/gtk/gtkdndprivate.h
+++ b/gtk/gtkdndprivate.h
@@ -56,6 +56,7 @@ void                    gtk_drag_set_icon_definition    (GdkDragContext
                                                          gint                    hot_y);
 void                    _gtk_drag_source_handle_event   (GtkWidget              *widget,
                                                          GdkEvent               *event);
+GdkModifierType         _gtk_drag_source_get_start_button_mask (GtkWidget       *widget);
 void                    _gtk_drag_dest_handle_event     (GtkWidget              *toplevel,
 				                         GdkEvent               *event);
 
diff --git a/gtk/gtkdragsource.c b/gtk/gtkdragsource.c
index bbe3d2e..62e621b 100644
--- a/gtk/gtkdragsource.c
+++ b/gtk/gtkdragsource.c
@@ -461,3 +461,12 @@ gtk_drag_source_set_icon_gicon (GtkWidget *widget,
   site->image_def = gtk_image_definition_new_gicon (icon);
 }
 
+GdkModifierType
+_gtk_drag_source_get_start_button_mask (GtkWidget *widget)
+{
+  GtkDragSourceSite *site;
+
+  site = g_object_get_data (G_OBJECT (widget), "gtk-site-data");
+
+  return site ? site->start_button_mask : 0;
+}
diff --git a/gtk/gtkiconview.c b/gtk/gtkiconview.c
index 735695e..ac34a95 100644
--- a/gtk/gtkiconview.c
+++ b/gtk/gtkiconview.c
@@ -984,11 +984,30 @@ gtk_icon_view_cell_layout_init (GtkCellLayoutIface *iface)
   iface->get_area = gtk_icon_view_cell_layout_get_area;
 }
 
+static gboolean
+gtk_icon_view_drag_source_hit_test (GtkWidget      *widget,
+                                    GdkEventButton *event)
+{
+  GtkIconView *icon_view = GTK_ICON_VIEW (widget);
+
+  if (event->window != icon_view->priv->bin_window)
+    return FALSE;
+
+  if (icon_view->priv->source_set &&
+      !(icon_view->priv->start_button_mask & (GDK_BUTTON1_MASK << (event->button - 1))))
+    return FALSE;
+
+  return _gtk_icon_view_get_item_at_coords (icon_view, event->x, event->y, TRUE, NULL) != NULL;
+}
+
 static void
 gtk_icon_view_init (GtkIconView *icon_view)
 {
   icon_view->priv = gtk_icon_view_get_instance_private (icon_view);
 
+  g_object_set_data (G_OBJECT (icon_view), I_("gtk-drag-source-hit-test"),
+                     (gpointer) gtk_icon_view_drag_source_hit_test);
+
   icon_view->priv->width = 0;
   icon_view->priv->height = 0;
   icon_view->priv->selection_mode = GTK_SELECTION_SINGLE;
diff --git a/gtk/gtkmain.c b/gtk/gtkmain.c
index 3c4a010..b02acee 100644
--- a/gtk/gtkmain.c
+++ b/gtk/gtkmain.c
@@ -131,6 +131,10 @@
 #include "gtkwindowprivate.h"
 #include "gtkwindowgroup.h"
 
+#ifdef GDK_WINDOWING_X11
+#include "x11/gdkx.h"
+#endif
+
 #include "a11y/gtkaccessibility.h"
 
 /* Private type definitions
@@ -1649,6 +1653,129 @@ check_event_in_child_popover (GtkWidget *event_widget,
   return (popover_parent == grab_widget || gtk_widget_is_ancestor (popover_parent, grab_widget));
 }
 
+#ifdef GDK_WINDOWING_X11
+typedef gboolean (*GtkDragSourceHitTestFunc) (GtkWidget      *widget,
+                                              GdkEventButton *event);
+
+static gboolean
+gtk_widget_chain_is_drag_source (GtkWidget      *event_widget,
+                                 GdkEventButton *event)
+{
+  static GQuark site_quark = 0;
+  static GQuark hit_test_quark = 0;
+  static guint drag_begin_id = 0;
+  static guint drag_data_get_id = 0;
+  GtkDragSourceHitTestFunc hit_test;
+  GdkModifierType start_button_mask;
+  GtkWidget *widget;
+
+  if (!site_quark)
+    {
+      site_quark = g_quark_from_static_string ("gtk-site-data");
+      hit_test_quark = g_quark_from_static_string ("gtk-drag-source-hit-test");
+      drag_begin_id = g_signal_lookup ("drag-begin", GTK_TYPE_WIDGET);
+      drag_data_get_id = g_signal_lookup ("drag-data-get", GTK_TYPE_WIDGET);
+    }
+
+  for (widget = event_widget; widget; widget = gtk_widget_get_parent (widget))
+    {
+      if (g_object_get_qdata (G_OBJECT (widget), site_quark) ||
+          g_signal_has_handler_pending (widget, drag_begin_id, 0, TRUE) ||
+          g_signal_has_handler_pending (widget, drag_data_get_id, 0, TRUE))
+        {
+          start_button_mask = _gtk_drag_source_get_start_button_mask (widget);
+          if (start_button_mask && !(start_button_mask & (GDK_BUTTON1_MASK << (event->button - 1))))
+            return FALSE;
+          hit_test = (GtkDragSourceHitTestFunc) g_object_get_qdata (G_OBJECT (widget), hit_test_quark);
+          return hit_test ? hit_test (widget, event) : TRUE;
+        }
+    }
+
+  return FALSE;
+}
+
+static Window activate_on_release_xid = None;
+static guint activate_on_release_button = 0;
+
+static void
+gtk_main_send_activate_on_click (GdkDisplay *display,
+                                 Window      xid,
+                                 guint32     time,
+                                 guint       button,
+                                 gboolean    on_release)
+{
+  XClientMessageEvent xev = { 0 };
+
+  xev.type = ClientMessage;
+  xev.window = xid;
+  xev.message_type = gdk_x11_get_xatom_by_name_for_display (display, "_GTK_ACTIVATE_ON_CLICK");
+  xev.format = 32;
+  xev.data.l[0] = time;
+  xev.data.l[1] = button;
+  xev.data.l[2] = on_release;
+  XSendEvent (GDK_DISPLAY_XDISPLAY (display),
+              DefaultRootWindow (GDK_DISPLAY_XDISPLAY (display)),
+              False, SubstructureRedirectMask, (XEvent *) &xev);
+  gdk_display_flush (display);
+}
+#endif
+
+void
+_gtk_main_cancel_activate_on_release (void)
+{
+#ifdef GDK_WINDOWING_X11
+  activate_on_release_xid = None;
+#endif
+}
+
+static void
+gtk_main_activate_on_click (GtkWidget      *event_widget,
+                            GdkEventButton *event)
+{
+#ifdef GDK_WINDOWING_X11
+  GdkWindow *toplevel;
+  gboolean on_release;
+
+  if (!GDK_IS_X11_WINDOW (event->window))
+    return;
+
+  if (event->type == GDK_BUTTON_RELEASE)
+    {
+      if (activate_on_release_xid && event->button == activate_on_release_button)
+        {
+          gtk_main_send_activate_on_click (gdk_window_get_display (event->window),
+                                           activate_on_release_xid, event->time, event->button, FALSE);
+          activate_on_release_xid = None;
+        }
+      return;
+    }
+
+  if (activate_on_release_xid &&
+      (event->state & (GDK_BUTTON1_MASK << (activate_on_release_button - 1))))
+    return;
+
+  activate_on_release_xid = None;
+
+  if (event->button > 3)
+    return;
+
+  toplevel = gdk_window_get_toplevel (event->window);
+  if (!gdk_x11_screen_supports_net_wm_hint (gdk_window_get_screen (toplevel),
+                                            gdk_atom_intern_static_string ("_GTK_ACTIVATE_ON_CLICK")))
+    return;
+
+  on_release = gtk_widget_chain_is_drag_source (event_widget, event);
+  if (on_release)
+    {
+      activate_on_release_xid = GDK_WINDOW_XID (toplevel);
+      activate_on_release_button = event->button;
+    }
+
+  gtk_main_send_activate_on_click (gdk_window_get_display (toplevel),
+                                   GDK_WINDOW_XID (toplevel), event->time, event->button, on_release);
+#endif
+}
+
 /**
  * gtk_main_do_event:
  * @event: An event to process (normally passed by GDK)
@@ -1720,6 +1847,10 @@ gtk_main_do_event (GdkEvent *event)
    * bogus events can occur for e.g. destroyed GdkWindows.
    */
   event_widget = gtk_get_event_widget (event);
+
+  if (event->type == GDK_BUTTON_PRESS || event->type == GDK_BUTTON_RELEASE)
+    gtk_main_activate_on_click (event_widget, &event->button);
+
   if (!event_widget)
     {
       /* To handle selection INCR transactions, we select
diff --git a/gtk/gtkprivate.h b/gtk/gtkprivate.h
index 40936b9..84a193b 100644
--- a/gtk/gtkprivate.h
+++ b/gtk/gtkprivate.h
@@ -91,6 +91,8 @@ gboolean        _gtk_propagate_captured_event  (GtkWidget       *widget,
                                                 GdkEvent        *event,
                                                 GtkWidget       *topmost);
 
+void            _gtk_main_cancel_activate_on_release (void);
+
 
 gdouble _gtk_get_slowdown (void);
 void    _gtk_set_slowdown (gdouble slowdown_factor);
diff --git a/gtk/gtktreeview.c b/gtk/gtktreeview.c
index 79bd8c6..f7a28e0 100644
--- a/gtk/gtktreeview.c
+++ b/gtk/gtktreeview.c
@@ -1815,6 +1815,24 @@ G_GNUC_END_IGNORE_DEPRECATIONS
   gtk_widget_class_set_css_name (widget_class, "treeview");
 }
 
+static TreeViewDragInfo *get_info (GtkTreeView *tree_view);
+
+static gboolean
+gtk_tree_view_drag_source_hit_test (GtkWidget      *widget,
+                                    GdkEventButton *event)
+{
+  GtkTreeView *tree_view = GTK_TREE_VIEW (widget);
+  TreeViewDragInfo *di = get_info (tree_view);
+
+  if (event->window != tree_view->priv->bin_window)
+    return FALSE;
+
+  if (di && di->source_set && !(di->start_button_mask & (GDK_BUTTON1_MASK << (event->button - 1))))
+    return FALSE;
+
+  return gtk_tree_view_get_path_at_pos (tree_view, event->x, event->y, NULL, NULL, NULL, NULL);
+}
+
 static void
 gtk_tree_view_init (GtkTreeView *tree_view)
 {
@@ -1823,6 +1841,9 @@ gtk_tree_view_init (GtkTreeView *tree_view)
 
   priv = tree_view->priv = gtk_tree_view_get_instance_private (tree_view);
 
+  g_object_set_data (G_OBJECT (tree_view), I_("gtk-drag-source-hit-test"),
+                     (gpointer) gtk_tree_view_drag_source_hit_test);
+
   gtk_widget_set_can_focus (GTK_WIDGET (tree_view), TRUE);
 
   priv->show_expanders = TRUE;
